File: 0020. Valid Parentheses/Solution.py
class Solution(object):
    def isValid(self, s):
        st = []
        mp = {')':'(', '}':'{', ']':'['}

        for ch in s:
            if ch in mp:
                if not st or st.pop() != mp[ch]:
                    return False
            else:
                st.append(ch)

        return not st


#alternate solution

# class Solution(object):
#     def isValid(self, s):
#         st = []

#         for e in s:
#             if e == '(' or e == '{' or e == '[':
#                 st.append(e)
#             else:
#                 if not st:
#                     return False

#                 c = st.pop()

#                 if (e == ')' and c != '(') or (e == '}' and c != '{') or (e == ']' and c != '['):
#                     return False

#         return len(st) == 0

File: 0020. Valid Parentheses/Solution.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st =  new Stack<>();

        for(int i=0;i<s.length();i++){
            char e = s.charAt(i);
            if(e=='('||e=='{'||e=='['){
                st.push(e);
            }

            else{
                if (st.isEmpty()) return false;

                char c = st.pop();

                if((e==')' && c!='(')||(e=='}' && c!='{')||(e==']' && c!='['))return false;
            }
        }
        return st.isEmpty();
    } 
}